UB in ACPI_TO_INTEGER , ACPI_OFFSET

It is bogus in C to subtract arbitrary pointers that do not point to the same object (or one beyond the end of the object), and the use of difference for these is not needed.  (C11 §6.5.6, constraint 9, p.93 of the last draft.)

C defines offsetof() in stddef.h since C89, but even if we are concerned about running on environments that do not support offsetof, it would be more correct to just cast the corresponding pointers to `uintptr_t` and skip the subtraction.  NULL == 0, regardless of bit representation.

I suggest (something along the lines of):

```
#define ACPI_TO_INTEGER(p) ((uintptr_t)(p))
#define ACPI_OFFSET(d, f) (offsetof(d, f))
```

I ran into this issue again when building recent FreeBSD (which contains acpica) with clang 13.0.0:

```
sys/contrib/dev/acpica/components/dispatcher/dsopcode.c:708:31: error: performing pointer subtraction with a null pointer has undefined behavior [-Werror,-Wnull-pointer-subtraction]
```

This is indeed due to the line in `sys/contrib/dev/acpica/components/dispatcher/dsopcode.c`:

```
    ObjDesc->Region.Address = ACPI_PTR_TO_PHYSADDR (Table);
```

with `ACPI_PTR_TO_PHYSADDR` expanding to `ACPI_TO_INTEGER` in `sys/contrib/dev/acpica/include/actypes.h`:
#
```
/* Pointer/Integer type conversions */

#define ACPI_TO_POINTER(i)              ACPI_CAST_PTR (void, (ACPI_SIZE) (i))
#define ACPI_TO_INTEGER(p)              ACPI_PTR_DIFF (p, (void *) 0)
#define ACPI_OFFSET(d, f)               ACPI_PTR_DIFF (&(((d *) 0)->f), (void *) 0)
#define ACPI_PHYSADDR_TO_PTR(i)         ACPI_TO_POINTER(i)
#define ACPI_PTR_TO_PHYSADDR(i)         ACPI_TO_INTEGER(i)
```

Looking through `actypes.h` there's already a type that sort-of covers `intptr_t`, which is `ACPI_NATIVE_INT`, and both `uintptr_t` and `size_t` are covered by `ACPI_SIZE`. So wouldn't it be easier to define `ACPI_TO_INTEGER` as:

```
#define ACPI_TO_INTEGER(p)              ((ACPI_SIZE) (p))
```

This won't work for the `ACPI_OFFSET` macro though. The only portable way is to use `offsetof`, but it may be that this isn't supported by older Microsoft compilers. They were notoriously bad at supporting anything resembling modern C standards. (Though they've greatly cleaned up their act nowadays. :) )

--- a/source/include/platform/acfreebsd.h
+++ b/source/include/platform/acfreebsd.h
@@ -166,6 +166,9 @@
 
 #define ACPI_UINTPTR_T      uintptr_t
 
+#define ACPI_TO_INTEGER(p)  ((uintptr_t)(p))
+#define ACPI_OFFSET(d, f)   __offsetof(d, f)
+
 #define ACPI_USE_DO_WHILE_0
 #define ACPI_USE_LOCAL_CACHE
 #define ACPI_USE_NATIVE_DIVIDE

--- a/source/include/platform/acfreebsd.h
+++ b/source/include/platform/acfreebsd.h
@@ -211,6 +211,7 @@
 
 #if __STDC_HOSTED__
 #include <ctype.h>
+#include <unistd.h>
 #endif
 
 #define ACPI_CAST_PTHREAD_T(pthread)    ((ACPI_THREAD_ID) ACPI_TO_INTEGER (pthread))

